### PR TITLE
Fix `array::IntoIter::fold` to use the optimized `Range::fold`

### PR DESCRIPTION
**Pull request by scottmcm**
It was using `Iterator::by_ref` in the implementation, which ended up pessimizing it enough that, for example, it didn't vectorize when we tried it in the <https://rust-lang.zulipchat.com/#narrow/stream/257879-project-portable-simd/topic/Reducing.20sum.20into.20wider.20types> conversation.

Demonstration that the codegen test doesn't pass on the current nightly: <https://rust.godbolt.org/z/Taxev5eMn>

### CODE REVIEW

**Comment by the8472** [comment]
Huh, interesting. It previously used `try_fold` instead to not go through the `impl<T: Iterator> Iterator for &mut T` blanket impl (#87431), which was then removed in #89774 by adding `#[inline]` on the blanket impl which was sufficient for the `vec::bench_flat_map_collect` benchmark.
But yeah, this should be fine too.

**Comment by scottmcm** [comment]
Friendly one-month review reminder ping, @joshtriplett.

File: library/core/src/iter/adapters/by_ref_sized.rs [hunk 1]
```diff
@@ -40,3 +40,32 @@ impl<I: Iterator> Iterator for ByRefSized<'_, I> {
         self.0.try_fold(init, f)
     }
 }
+
+impl<I: DoubleEndedIterator> DoubleEndedIterator for ByRefSized<'_, I> {
+    fn next_back(&mut self) -> Option<Self::Item> {
```
Comment (the8472) [inline library/core/src/iter/adapters/by_ref_sized.rs:45]: It's probably better to slap `#[inline]` on these methods to ensure that the whole iterator gets inlined when compiling with many CGUs.

**Pull request merged**